drivers/sensor/grow_r502a/grow_r502a.c

```diff
@@ -0,0 +1,726 @@
+/*
+ * Copyright (c) 2021 Linumiz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT hzgrow_r502a
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/byteorder.h>
+
+#include "grow_r502a.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(GROW_R502A, CONFIG_SENSOR_LOG_LEVEL);
+
+static void send_packet(const struct device *dev, uint8_t pkg_type, uint8_t *pkg_data, uint16_t len)
+{
+	const struct grow_r502a_config *cfg = dev->config;
+	struct grow_r502a_data *drv_data = dev->data;
+	uint16_t check_sum, pkg_len;
+
+	k_mutex_lock(&drv_data->lock, K_FOREVER);
+
+	pkg_len = len + R502A_CHECKSUM_LEN;
+	check_sum = pkg_len + pkg_type;
+
+	sys_put_be16(R502A_STARTCODE, &drv_data->tx_buf.data[R502A_STARTCODE_IDX]);
+	sys_put_be32(cfg->comm_addr, &drv_data->tx_buf.data[R502A_ADDRESS_IDX]);
+	drv_data->tx_buf.data[R502A_PID_IDX] = pkg_type;
+	sys_put_be16(pkg_len, &drv_data->tx_buf.data[R502A_PKG_LEN_IDX]);
+	memcpy(&drv_data->tx_buf.data[R502A_HEADER_LEN], pkg_data, len);
+	for (int i = 0; i < len; i++) {
+		check_sum += pkg_data[i];
+	}
+	sys_put_be16(check_sum, &drv_data->tx_buf.data[len + R502A_HEADER_LEN]);
+
+	drv_data->tx_buf.len = pkg_len + R502A_HEADER_LEN;
+
+	LOG_HEXDUMP_DBG(drv_data->tx_buf.data, drv_data->tx_buf.len, "TX");
+
+	uart_irq_rx_disable(cfg->dev);
+	uart_irq_tx_enable(cfg->dev);
+
+	k_sem_take(&drv_data->uart_rx_sem, K_FOREVER);
+	k_mutex_unlock(&drv_data->lock);
+}
+
+static void uart_cb_tx_handler(const struct device *dev)
+{
+	const struct grow_r502a_config *config = dev->config;
+	struct grow_r502a_data *drv_data = dev->data;
+	int sent;
+
+	while (drv_data->tx_buf.len) {
+		sent = uart_fifo_fill(config->dev, drv_data->tx_buf.data, drv_data->tx_buf.len);
+		drv_data->tx_buf.len -= (uint8_t)sent;
+	}
+
+	if (uart_irq_tx_complete(config->dev) > 0) {
+		uart_irq_tx_disable(config->dev);
+		uart_irq_rx_enable(config->dev);
+	}
+}
+
+static void uart_cb_handler(const struct device *dev, void *user_data)
+{
+	const struct device *uart_dev = user_data;
+	struct grow_r502a_data *drv_data = uart_dev->data;
+
+	if ((uart_irq_update(dev) > 0) && (uart_irq_is_pending(dev) > 0)) {
+		while (uart_irq_rx_ready(dev)) {
+			drv_data->rx_buf.len = uart_fifo_read(dev, drv_data->rx_buf.data,
```

Comment:
read in rx_buf.data index as rx_buf.len and use local variable for number of byte read in fifo_read, then update rx_buf.len. if full packet is not read , then continue

Reply:
`full packet is not read` but packet length will vary for each tx data packet, so do you think it is better to calculate the expected reception data length, this will be useful in checking full packet is received or not